### PR TITLE
Propagate delete companion

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [na] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR addresses https://github.com/ibi-group/otp-middleware/pull/271#discussion_r1860433391.

When a user deletes their account, they are removed as primary traveler on trips saved by others,
and they are invalidated as companions and observers on trips saved by others.

### CODE REVIEW

File: src/main/java/org/opentripplanner/middleware/models/OtpUser.java [hunk 1]
```diff
@@ -143,20 +148,34 @@ public boolean delete(boolean deleteAuth0User) {
...
 
-        // If a dependent, remove relationship with all related users.
+        // If a dependent user, remove relationship with all related users.
         for (RelatedUser relatedUser : relatedUsers) {
             removeDependent(this, relatedUser);
         }
+
+        // If a dependent user, remove self as the primary traveler in trips created by other users.
+        // TODO: Should we alert the user who created the trip of the deletion?
+        Persistence.monitoredTrips
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/models/OtpUser.java:164]: There is probably a case here for just deleting the trip if the primary traveler is no longer travelling. Perhaps that would start to complicate things though?!

File: src/main/java/org/opentripplanner/middleware/tripmonitor/TrustedCompanion.java [hunk 2]
```diff
@@ -248,6 +250,62 @@ public static void removeDependent(OtpUser dependent, RelatedUser relatedUser) {
...
+        ) {
+            relatedUser.status = RelatedUser.RelatedUserStatus.INVALID;
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Remove the specified user as observer from the specified trip.
+     * @return true if one or more related users gor invalidated, false otherwise.
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/tripmonitor/TrustedCompanion.java]: Nit. `are` not `gor`.

**Review by br648 — APPROVED** [review]
LGTM.

File: src/main/java/org/opentripplanner/middleware/tripmonitor/TrustedCompanion.java (hunk 2 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/tripmonitor/TrustedCompanion.java]: Thanks! https://github.com/ibi-group/otp-middleware/pull/278/commits/528d8bbb2e4bbbb431752b5d6e418f96e62ccf34

File: src/main/java/org/opentripplanner/middleware/models/OtpUser.java (hunk 1 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/OtpUser.java:164]: Good observation. Could preventing tracking if a companion is set but no primary traveler is defined be a better way to handle that?

File: src/main/java/org/opentripplanner/middleware/models/OtpUser.java [hunk 3]
```diff
@@ -143,20 +148,34 @@ public boolean delete(boolean deleteAuth0User) {
...
+            .getFiltered(Filters.eq("primary.userId", id))
+            .forEach(trip -> removePrimaryTraveler(this, trip));
+
+        // If a companion user, invalidate relationship in trips where they are companions and observers.
+        // TODO: Should we alert the user who created the trip of the deletion?
+        Persistence.monitoredTrips
+            .getFiltered(Filters.eq("companion.email", email))
+            .forEach(trip -> removeCompanion(this, trip));
+
+        // If a companion user, invalidate relationship in trips where they are companions and observers.
```
Comment (josh-willis-arcadis) [inline src/main/java/org/opentripplanner/middleware/models/OtpUser.java]: nit. this comment is a duplicate of the comment above.

File: src/main/java/org/opentripplanner/middleware/models/OtpUser.java (hunk 1 above)
Comment (josh-willis-arcadis) [reply src/main/java/org/opentripplanner/middleware/models/OtpUser.java:164]: I think deleting the trip makes sense if there is no primary traveler.

File: src/main/java/org/opentripplanner/middleware/models/OtpUser.java [hunk 4]
```diff
@@ -143,20 +148,34 @@ public boolean delete(boolean deleteAuth0User) {
...
+        // TODO: Should we alert the user who created the trip of the deletion?
```
Comment (josh-willis-arcadis) [inline src/main/java/org/opentripplanner/middleware/models/OtpUser.java]: I don't think it would hurt to notify the creator of the trip.

**Review by josh-willis-arcadis — APPROVED** [review]
LGTM. Just one tiny nit

File: src/main/java/org/opentripplanner/middleware/models/OtpUser.java (hunk 1 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/OtpUser.java:164]: Created issue #282 to track that.

File: src/main/java/org/opentripplanner/middleware/models/OtpUser.java (hunk 4 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/OtpUser.java]: Created issue #283 to track that. The dilemma is to choose between #282 and #283.

**Pull request merged**